### ISSUE TITLE
tensorflow codes totally different result with PyTorch's

### ISSUE THREAD

**Issue by Response777** [tensorflow/tensorflow#7624]
Hi, I'm new to tensorflow and trying to refactor one of my project originally in (Py)Torch. However, though the two codes has the same network, the same loss function, the same optimizer parameters, the result is totally different. Is any one have any idea about such issue?

Thanks.

**Comment by yaroslavvb** [comment]
This question may be better suited for stackoverflow. In general there's no magic solution short of comparing results line by line -- because there's no standard, neural network primitives have slightly different implementations across frameworks

**Comment by Response777** [comment]
Thanks. However, the result is stable. In fact, I'm training a conditional VAE, and for the  tensorflow model, the category label hardly influence the image generated while the PyTorch's run smoothly. So it is likely that something is going wrong.

Currently, I'm trying to test the model with an older version of tensorflow, and to check is this would only occur in tensorflow 1.0.

**Comment by yaroslavvb** [comment]
If you are able to isolate it to a specific operation that does an incorrect result, it would be worth filing an issue with reproduction instructions. I've seen examples of networks not training because of some slight differences in op semantics

**Comment by aselle** [comment]
In addition to @yaroslavvb suggesting we need a reproducible test case, please provide details about what platform you are using  (operating system, architecture). Also include your TensorFlow version. We ask for this in the issue submission template, because    it is really difficult to help without that information. Thanks! You could also try with and without GPU.

**Comment by Response777** [comment]
The code can be viewed at https://github.com/Response777/Tensorflow-VAEs.

The tensorflow code's generated result of conditional VAE is in [Tensorflow-VAEs//misc/condition_generated_imgs/cvae-mlp-mnist/](https://github.com/Response777/Tensorflow-VAEs/tree/master/misc/condition_generated_imgs/cvae-mlp-mnist)

While the PyTorch's is in [Tensorflow-VAEs/PyTorch/CVAE.py/](https://github.com/Response777/Tensorflow-VAEs/tree/master/PyTorch/CVAE.py)

It seems that the label can hardly influence the generated images, while in PyTorch, the code just runs smoothly.

**Comment by Response777** [comment]
The PyTorch and Tensorflow network shares the same network structure.
And I use the same loss function, the same Optimizer, the same learning_rate.

And I have checked the weights initializer of tensorflow.contrib.layers.fully_connected and PyTorch's torch.nn.modules.linear. 

Now, the tensorflow's layer is using variance_scaling_initializer(factor = 0.3, mode = 'FAN_IN', uniform = True), which is the same as the one in PyTorch's(this could be examined at http://pytorch.org/docs/_modules/torch/nn/modules/linear.html#Linear)

**Comment by yaroslavvb** [comment]
Can you isolate the difference to a specific op? (Ie it's not clear if this
is caused by a bug, or by difference in op semantics which is not a bug)

**Comment by aselle** [comment]
Also, if you are using random initialization, you might want to switch to constant initialization, followed by seeded initialization.

**Comment by Response777** [comment]
@yaroslavvb It seems that the results are different even if I kept all the conditions the same, so I'm having difficulty control a specific op, since PyTorch and Tensorflow are two distinct framework.

Or you means I should not use the layer in **tensorflow.contrib.layers**? If so, I'm trying it at the same time, but coming across with a new problem...

**Comment by Response777** [comment]
@aselle It seems that I need to see if the PyTorch and Tensorflow are using the same random function first, or it won't help even if the seed is the same.

**Comment by Response777** [comment]
@aselle I found that the PyTorch's `uniform_(from=0, to=1) → Tensor` does not have seed as an input

**Comment by Response777** [comment]
@yaroslavvb I'm trying to isolate the difference to an op by writing layers ([module/mnist/cvae/mlp2.py](https://github.com/Response777/Tensorflow-VAEs/blob/master/modules/mnist/cvae/mlp2.py)).

The training procedure runs smoothly, but the sampling procedure is coming across with some problem: they are not sharing variables, even if I have added

```
with tf.variable_scope(self.name) as vs:
            if reuse == True:
                vs.reuse_variables()
```

According to the output of `tf.all_variables()`, the program have created another decoder named `CVAE/Decoder_1`, but I have set `reuse = True` in `CVAE::__init__`, and the code didn't throw any exception

**Comment by Response777** [comment]
Fixed.

**Comment by JingHuang81** [comment]
@Response777 What was your problem with fully_connected layer and how did you fix it? I was transferring a model from PyTroch to TensorFlow and observed a decrease in performance. I also used the tf.contrib.layers.fully_connected to replace the Pytorch nn.Linear, so I wonder if I ran into the same issue.

**Comment by Response777** [comment]
Hi @JingHuang81 , I believe they are two distinct issues -- I was training new model from scratch at that time rather than just transfering the parameters. As far as I am concerned, your problem might be related to some differences in implementation or numerical unstablility. I think you might want to  compare intermediate results layer by layer(or you can try binary search to locate the causes).

**Comment by vanpersie32** [comment]
@Response777 Hi, can you please tell me what was your problem with fully_connected layer. I encounter a performance drop when I convert a torch version model to a tensorflow version model.
